File: src/utils/url-utils.js
export function normalizeUrl(rawUrl, options = {}) {
  const {
    removeHash = true,
    removeTrailingSlash = false,
    removeCommonTrackingParams = true,
    trackingParams = []
  } = options;

  let parsed;

  try {
    parsed = new URL(rawUrl);
  } catch (error) {
    throw new Error(`Invalid URL: ${rawUrl}`);
  }

  parsed.protocol = parsed.protocol.toLowerCase();
  parsed.hostname = parsed.hostname.toLowerCase();

  if (removeHash) {
    parsed.hash = '';
  }

  if (removeCommonTrackingParams) {
    for (const param of trackingParams) {
      parsed.searchParams.delete(param);
    }
  }

  const sortedParams = [...parsed.searchParams.entries()].sort(([a], [b]) =>
    a.localeCompare(b)
  );

  parsed.search = '';

  for (const [key, value] of sortedParams) {
    parsed.searchParams.append(key, value);
  }

  let normalized = parsed.toString();

  if (removeTrailingSlash) {
    normalized = normalized.replace(/\/$/, '');
  }

  return normalized;
}

export function safeNormalizeUrl(rawUrl, options = {}) {
  try {
    return normalizeUrl(rawUrl, options);
  } catch {
    return null;
  }
}

export function slugify(input) {
  return String(input)
    .normalize('NFKD')
    .replace(/[^\w\s-]/g, '')
    .trim()
    .replace(/\s+/g, '_')
    .replace(/_+/g, '_')
    .toLowerCase();
}

export function deduplicatePages(pages, normalizationOptions = {}) {
  const seen = new Set();
  const uniquePages = [];
  const duplicates = [];

  for (const page of pages) {
    const normalizedUrl = normalizeUrl(page.url, normalizationOptions);

    const enrichedPage = {
      ...page,
      normalizedUrl
    };

    if (seen.has(normalizedUrl)) {
      duplicates.push(enrichedPage);
      continue;
    }

    seen.add(normalizedUrl);
    uniquePages.push(enrichedPage);
  }

  return {
    uniquePages,
    duplicates
  };
}

export function getOriginSafe(rawUrl) {
  try {
    return new URL(rawUrl).origin;
  } catch {
    return null;
  }
}

File: src/utils/url_utils.py
import re
import unicodedata
from urllib.parse import parse_qsl, urlsplit, urlunsplit, urlencode


WINDOWS_RESERVED_NAMES = {
    "CON",
    "PRN",
    "AUX",
    "NUL",
    "COM1",
    "COM2",
    "COM3",
    "COM4",
    "COM5",
    "COM6",
    "COM7",
    "COM8",
    "COM9",
    "LPT1",
    "LPT2",
    "LPT3",
    "LPT4",
    "LPT5",
    "LPT6",
    "LPT7",
    "LPT8",
    "LPT9",
}


def normalize_url(raw_url, options=None):
    config = options or {}
    remove_hash = config.get("removeHash", True)
    remove_trailing_slash = config.get("removeTrailingSlash", False)
    remove_common_tracking_params = config.get("removeCommonTrackingParams", True)
    tracking_params = set(config.get("trackingParams", []))

    try:
        parsed = urlsplit(raw_url)
    except Exception as error:
        raise ValueError(f"Invalid URL: {raw_url}") from error

    if not parsed.scheme or not parsed.netloc:
        raise ValueError(f"Invalid URL: {raw_url}")

    query_items = parse_qsl(parsed.query, keep_blank_values=True)
    if remove_common_tracking_params:
        query_items = [(key, value) for key, value in query_items if key not in tracking_params]

    query_items.sort(key=lambda item: item[0])
    fragment = "" if remove_hash else parsed.fragment

    normalized = urlunsplit(
        (
            parsed.scheme.lower(),
            parsed.netloc.lower(),
            parsed.path,
            urlencode(query_items, doseq=True),
            fragment,
        )
    )

    if remove_trailing_slash:
        normalized = re.sub(r"/$", "", normalized)

    return normalized


def safe_normalize_url(raw_url, options=None):
    try:
        return normalize_url(raw_url, options)
    except Exception:
        return None


def slugify(value):
    normalized = unicodedata.normalize("NFKD", str(value or ""))
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    slug = re.sub(r"[^\w\s-]", "", ascii_text).strip()
    slug = re.sub(r"\s+", "_", slug)
    slug = re.sub(r"_+", "_", slug)
    return slug.lower()


def deduplicate_pages(pages, normalization_options=None):
    seen = set()
    unique_pages = []
    duplicates = []

    for page in pages:
        normalized_url = normalize_url(page["url"], normalization_options or {})
        enriched_page = {
            **page,
            "normalizedUrl": normalized_url,
        }

        if normalized_url in seen:
            duplicates.append(enriched_page)
            continue

        seen.add(normalized_url)
        unique_pages.append(enriched_page)

    return {
        "uniquePages": unique_pages,
        "duplicates": duplicates,
    }


def get_origin_safe(raw_url):
    try:
        parsed = urlsplit(raw_url)
        if not parsed.scheme or not parsed.netloc:
            return None
        return f"{parsed.scheme}://{parsed.netloc}"
    except Exception:
        return None


def get_hostname_safe(raw_url):
    try:
        parsed = urlsplit(raw_url)
        return parsed.hostname.lower() if parsed.hostname else None
    except Exception:
        return None


def sanitize_path_segment(value, fallback="item", keep_dots=False):
    normalized = unicodedata.normalize("NFKD", str(value or ""))
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    cleaned = re.sub(r'[<>:"/\\|?*]', " ", ascii_text)

    if not keep_dots:
        cleaned = cleaned.replace(".", " ")

    cleaned = re.sub(r"\s+", " ", cleaned).strip(" .")

    if not cleaned:
        cleaned = fallback

    if cleaned.upper() in WINDOWS_RESERVED_NAMES:
        cleaned = f"_{cleaned}"

    return cleaned[:100]


def build_website_folder_name(raw_url):
    hostname = get_hostname_safe(raw_url)
    return sanitize_path_segment(hostname or "website", fallback="website", keep_dots=True)


def build_page_folder_name(page_name, fallback="page"):
    return sanitize_path_segment(page_name or fallback, fallback=fallback, keep_dots=False)
